### PR TITLE
Added the block ip action.

### PR DESCRIPTION
**Pull request by chindris**
Port "Block ip" action to D8: https://www.drupal.org/node/2317205

### CODE REVIEW

File: src/Plugin/Action/BlockIP.php [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+ */
+class BlockIP extends RulesActionBase implements ContainerFactoryPluginInterface {
+
+  /**
+   * @var \Drupal\ban\BanIpManagerInterface $banManger
+   */
+  protected $banManager;
+
+  /**
+   * @param \Drupal\ban\BanIpManagerInterface $banManager
```
Comment (fago) [inline src/Plugin/Action/BlockIP.php]: Incomplete docs.

File: src/Plugin/Action/BlockIP.php [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+   */
+  public function __construct(array $configuration, $plugin_id, $plugin_definition, BanIpManagerInterface $banManager) {
+    parent::__construct($configuration, $plugin_id, $plugin_definition);
+    $this->banManager = $banManager;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container, array $configuration, $plugin_id, $plugin_definition) {
```
Comment (fago) [inline src/Plugin/Action/BlockIP.php:48]: Minor, but ordering would be logical to have that on the top.

File: src/Plugin/Action/BlockIP.php [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+   */
+  public function summary() {
+    return $this->t('Blocks an IP address');
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function execute() {
+    $ip = String::checkPlain($this->getContextValue('ip'));
```
Comment (fago) [inline src/Plugin/Action/BlockIP.php]: Why do we need check plain to ban an ip? Seems to be not in d7 either.

File: src/Plugin/Action/BlockIP.php [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    $this->banManager->banIp($ip);
```
Comment (fago) [inline src/Plugin/Action/BlockIP.php:86]: This misses the watchdog call from d7.

File: src/Plugin/Action/BlockIP.php [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+ *   context = {
+ *     "ip" = @ContextDefinition("string",
+ *       label = @Translation("IP Address")
+ *     )
+ *   }
+ * )
+ *
+ * @todo: Add access callback information from Drupal 7.
+ * @todo: We should maybe use a dedicated data type for the ip address, as we
+ * do in Drupal 7.
```
Comment (fago) [inline src/Plugin/Action/BlockIP.php]: Also, this misses a dependency on ban module. We do not have a way to specify that yet, but let's add a todo for it.

**Comment by chindris** [comment]
I fixed the issues, one thing I did not do yet, that watchdog entry. Wouldn't make more sense that the watchdog entry is created by the ban module and not by the rules module?

File: src/Plugin/Action/BlockIP.php [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+ * Provides the 'Block IP' action.
+ *
+ * @Action(
+ *   id = "rules_block_ip",
+ *   label = @Translation("Block IP"),
+ *   category = @Translation("System"),
+ *   context = {
+ *     "ip" = @ContextDefinition("string",
+ *       label = @Translation("IP Address")
+ *     )
```
Comment (klausi) [inline src/Plugin/Action/BlockIP.php:26]: In D7 the IP parameter was optional and if not provided the IP address of the current user would be used. Ithink we should do the same here again.

Comment (fago) [inline src/Plugin/Action/BlockIP.php:26]: Exactly, it should be optional - just make it "required" => false.
It should have a description that describes the fallback also.

File: src/Plugin/Action/BlockIP.php [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+ * @todo: Add access callback information from Drupal 7.
+ * @todo: We should maybe use a dedicated data type for the ip address, as we
+ * do in Drupal 7.
+ * @todo: This action depends on the ban module. We need to have a way to
+ * specify this.
+ */
+class BlockIP extends RulesActionBase implements ContainerFactoryPluginInterface {
+
+  /**
+   * The ban manager service used to ban the IP.
```
Comment (fago) [inline src/Plugin/Action/BlockIP.php:39]: Minor, but the class does not know it's a service. So it's just The ban manager.

File: src/Plugin/Action/BlockIP.php [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+   * Constructs the BlockIP object.
+   *
+   * @param array $configuration
+   *   A configuration array containing information about the plugin instance.
+   * @param string $plugin_id
+   *   The plugin ID for the plugin instance.
+   * @param mixed $plugin_definition
+   *   The plugin implementation definition.
+   * @param \Drupal\ban\BanIpManagerInterface $banManager
+   *   The ban manager service.
```
Comment (fago) [inline src/Plugin/Action/BlockIP.php:67]: same here.

File: tests/src/Integration/Action/BlockIPTest.php [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+   */
+  protected $banManager;
+
+  /**
+   * {@inheritdoc}
+   */
+  public function setUp() {
+    parent::setUp();
+    // We need the ban module.
+    $this->enableModule('ban');
```
Comment (fago) [inline tests/src/Integration/Action/BlockIPTest.php:36]: Do we? It looks like the class is mocked below, so it should be fine without.

File: tests/src/Integration/Action/BlockIPTest.php [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+   */
+  public function testActionExecution() {
+    $ip = '192.155.122.111';
+    $this->action->setContextValue('ip', $ip);
+
+    $this->banManager->expects($this->once())
+      ->method('banIp')
+      ->with($ip);
+
+    $this->action->execute();
```
Comment (fago) [inline tests/src/Integration/Action/BlockIPTest.php:65]: Misses test coverage for the case where it takes the current IP as default.

File: src/Plugin/Action/BlockIP.php [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+  public function summary() {
+    return $this->t('Blocks an IP address');
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function execute() {
+    $ip = $this->getContextValue('ip');
+    $this->banManager->banIp($ip);
```
Comment (fago) [inline src/Plugin/Action/BlockIP.php:86]: Is the fallback to the current ip in ban manager?

**Comment by dasjo** [comment]
There is a follow-up
https://github.com/fago/rules/pull/142

**Pull request closed without merging**